BUILD-186: move off use of ObjectReference type in BuildConfigStatus.ImageChangeTriggers; remove Paused field

gabemontero opened the pull request:
Per feedback from @deads2k while reviewing https://github.com/openshift/openshift-apiserver/pull/191
the API originally introduced for 4.8 feature [BUILD-186](https://issues.redhat.com/browse/BUILD-186) via https://github.com/openshift/api/pull/855 
needs to move off of the use of ObjectReference.

Also removed the Paused field BuildConfigStatus.ImageChangeTriggers.  The existing field in the Spec is sufficient.

Also added clarifications in the godoc based on questions posed in @deads2k 's review.

/assign @deads2k 
/assign @adambkaplan 

@bparees FYI

deads2k commented on build/v1/types.go:
```diff
@@ -1022,23 +1026,37 @@ type ImageChangeTrigger struct {
+// generic ObjectReference BuildTriggerPolicy.ImageChange.From.  It is used by ImageChangeTriggerStatus, where a
+// specific instance of ImageChangeTriggerStatus in maintained in BuildConfigStatus.ImageChangeTriggers for each entry
+// in the BuildConfigSpec.Triggers array where the BuildTriggerPolicy.ImageChange pointer is set to a non-nil value
+type BuildImageChangeTriggerKey struct {
```
`ImageStreamTagReference`?

It's legal to have BuildConfig in ns/one triggered by an imagestreamtag in ns/two?

bparees replied:
> It's legal to have BuildConfig in ns/one triggered by an imagestreamtag in ns/two?

yes.  common, even.  "i want my app image in my NS to be rebuilt every time the corp base image in some shared NS it is built on top of is updated"

gabemontero replied:
Changed to `ImageStreamTagReference` in next push

deads2k commented on build/v1/types.go:
```diff
@@ -977,7 +977,11 @@ type BuildConfigStatus struct {
 
 	// ImageChangeTriggers is used to capture the runtime state of any ImageChangeTrigger specified in the BuildConfigSpec,
-	// including reconciled values of the lastTriggeredImageID and paused fields.
+	// including the value reconciled by the OpenShift APIServer for the lastTriggeredImageID.  There will be single entry
```
this is suggesting that the openshift-apiserver is setting a status field?  That is pretty unusual.  What causes the openshift-apiserver to set this?

deads2k replied:
> this is suggesting that the openshift-apiserver is setting a status field? That is pretty unusual. What causes the openshift-apiserver to set this?

if the openshift-apiserver is setting this value, can you describe which resource or subresource you're hitting that forces this calculation?

bparees replied:
iirc, a buildrequest comes into the apiserver.  The apiserver processes that buildrequest against the buildconfig to create a new build object.  As part of that, it must reconcile the current SHA values.

Once the apiserver has created the build object, it updates the buildconfig w/ information about the most recent build object it created.

Why buildrequests are fully processed in the apiserver(assuming my memory is correct about how it works) as part of the request flow and not in a controller is probably an artifact of the time period in which the code was written.  It'd probably be a controller loop today.

gabemontero replied:
@bparees 's account of the flow, including the fact the original time period on when the feature was introduced influencing the implementation then, and how things have been left untouched since then, are what I discovered while doing this feature.

I've certainly applied "minimal change needed" to my mindset while working on this change.  But of course if we reach a consensus that we want to take on additional changes to further "modernize", I'll react accordingly.  I have no strong opinion wrt that.

I'll put some code links and historical PR refs for reference.  I'll use existing merged code vs. what I have in https://github.com/openshift/openshift-apiserver/pull/191

1) ICTs for Builds was introduced in Aug 2015 :-) where the setting of the `LastTriggeredImageID` field in the apiserver first occurred at https://github.com/openshift/origin/blob/603eab741040fbb06219de02587ca6dce45837b2/pkg/build/generator/generator.go#L226 .... if you look at the `updateImageTriggers` back then, it does not look too different from today at https://github.com/openshift/openshift-apiserver/blob/7c32a64c6f1ea0d7a3e8f96bce0232d48c568eac/pkg/build/apiserver/buildgenerator/generator.go#L342-L382

2) Similarly, the flow that leads to the call to `updateImageTrigger` have remained similar.  Back in Aug 2015 it occurred in the `Instantiate` method.  See https://github.com/openshift/origin/blob/603eab741040fbb06219de02587ca6dce45837b2/pkg/build/generator/generator.go#L168
Then in Instantiate, the client `UpdateBuildConfig` was called.

The same occurs today, see https://github.com/openshift/openshift-apiserver/blob/7c32a64c6f1ea0d7a3e8f96bce0232d48c568eac/pkg/build/apiserver/buildgenerator/generator.go#L251 from `instantiate` for the call to `updateImageTriggers`, followed by the call the the client `UpdateBuildConfig` at https://github.com/openshift/openshift-apiserver/blob/7c32a64c6f1ea0d7a3e8f96bce0232d48c568eac/pkg/build/apiserver/buildgenerator/generator.go#L316

3) that `UpdateBuildConfig` in today's code ultimately translates to this REST invocation:  https://github.com/openshift/openshift-apiserver/blob/f85871147bb7c03c544fbc9c6515048ec470844d/vendor/github.com/openshift/client-go/build/clientset/versioned/typed/build/v1/buildconfig.go#L112-L123

deads2k commented on build/v1/types.go:
```diff
@@ -1022,23 +1026,37 @@ type ImageChangeTrigger struct {
 // specified in the BuildConfigSpec.Triggers struct.
 type ImageChangeTriggerStatus struct {
-	// lastTriggeredImageID is the sha/id of the imageref cited in the 'from' field the last time this BuildConfig was triggered.
+	// lastTriggeredImageID represents the current sha/id of the ImageStreamTag cited in the 'from' field of a corresponding
```
current from what perspective?  Last one used when the new build was created somehow?

bparees replied:
yes, the value at the time the buildconfig most recently instantiated a build.

@gabemontero the current text isn't quite correct if i recall the behavior:

> the last time a Build for this BuildConfig was instantiated as a result of *ANY* of the ImageChangeTriggers
in the BuildConfigSpec.Triggers array firing

it's not just when a build was instantiated as a result of a trigger firing.  A webhook, manual invocation, commit hook, etc all have the same behavior.  Any time the buildconfig runs a build, these imageids all get updated to their current values (reflecting the SHA value they had at the time the last build was run, and ensuring we don't immediately run a 2nd build for the same SHA)

gabemontero replied:
OK I will add the "instantiate in any way not just ICT" element @bparees 

I was trying to convey the `current from what perspective` element @deads2k asked about ... I'll take another stab at making that clearer.

Both in the next push.

deads2k commented on build/v1/types.go:
```diff
@@ -1022,23 +1026,37 @@ type ImageChangeTrigger struct {
-	// lastTriggeredImageID is the sha/id of the imageref cited in the 'from' field the last time this BuildConfig was triggered.
+	// lastTriggeredImageID represents the current sha/id of the ImageStreamTag cited in the 'from' field of a corresponding
+	// ImageChangeTrigger, captured by the non-nil BuildTriggerPolicy.ImageChange pointer from a BuildConfigSpec.Triggers
+	// entry, the last time a Build for this BuildConfig was instantiated as a result of *ANY* of the ImageChangeTriggers
```
This seems to suggest the value isn't current, but actually a snapshot of the imageID the last time a build was instantiated. If it is instantiation, please list the subresource to use.

gabemontero replied:
OK I'm attempting to do this in the next push as well.  Though I'm not super confident I'm correctly translating your request into the godoc.  We'll find out soon enough.

deads2k commented on build/v1/types.go:
```diff
@@ -1022,23 +1026,37 @@ type ImageChangeTrigger struct {
 
-	// lastTriggerTime is the last time the BuildConfig was triggered by a change in the ImageStreamTag associated with this trigger.
-	LastTriggerTime metav1.Time `json:"lastTriggerTime,omitempty" protobuf:"bytes,4,opt,name=lastTriggerTime"`
+	// lastTriggerTime is the last time this particular ImageChangeTrigger fired and resulted in a Build being instantiated
```
When combined with `lastTriggeredImageID` does this mean that the `lastTriggerTime` is the same for every entry in this slice?  Or does it mean that `lastTriggerTime` doesn't actually match the time `lastTriggeredImageID` value?

bparees replied:
lastTriggerTime == time at which *this* trigger lead to a build being run
lastTriggerImageID == the image id resolved by this trigger the last time the build was run for any reason  (probably not a great name, but trying to keep the values aligned w/ what we already have in spec).  The main purpose of this value is so we can determine "have we run a build since the last time the imagetag's SHA changed?  If not, we need to run one now and then update this field"

gabemontero replied:
lastly attempting to clarify this as well in next push

gabemontero commented:
ok @bparees @deads2k godoc updates in response to your set of comments up to this point have been pushed

aside from reaching consensus on godoc verbiage, I'm assuming we need consensus on what to do with the old form of the apiserver updating bc.spec.trigger[i].ict.lastTriggeredImageID via `UpdateBuildConfig` is something we want to move away from (and update the godoc accordingly) after this merges and I move onto the apiserver and OCM PRs

bparees commented:
> I'm assuming we need consensus on what to do with the old form of the apiserver updating bc.spec.trigger[i].ict.lastTriggeredImageID via UpdateBuildConfig is something we want to move away from (and update the godoc accordingly) after this merges and I move onto the apiserver and OCM PRs

"if it ain't broke, don't fix it" is my view.  The fact that that code has survived largely unchanged for 6 years says to me it's not something that needs to be rewritten.  Especially as we look to shipwright as the future of builds.

gabemontero commented:
> > I'm assuming we need consensus on what to do with the old form of the apiserver updating bc.spec.trigger[i].ict.lastTriggeredImageID via UpdateBuildConfig is something we want to move away from (and update the godoc accordingly) after this merges and I move onto the apiserver and OCM PRs
> 
> "if it ain't broke, don't fix it" is my view. The fact that that code has survived largely unchanged for 6 years says to me it's not something that needs to be rewritten. Especially as we look to shipwright as the future of builds.

OK @bparees @adambkaplan @deads2k I've attempted to collect some due diligence investigation data around the discussion thread @deads2k started at https://github.com/openshift/api/pull/882#discussion_r607203586 on the fact that updates to the BuildConfig are made by the apiserver, and what moving those updates from the apiserver to the OCM would entail.

My attempt to summarize what I came up with:

Option 1:
Replacing the call to Update on the BuildConfig client at https://github.com/openshift/openshift-apiserver/blob/7c32a64c6f1ea0d7a3e8f96bce0232d48c568eac/pkg/build/apiserver/buildgenerator/generator.go#L316 can easily be changed to `UpdateStatus`.  

That still handles the incrementing of `bc.Status.LastVersion` (i.e. https://github.com/openshift/openshift-apiserver/blob/f85871147bb7c03c544fbc9c6515048ec470844d/vendor/github.com/openshift/api/build/v1/types.go#L964) that is used to generate the next build name when creating builds off of the BuildConfig Instantiation endpoint (i.e. https://github.com/openshift/openshift-apiserver/blob/f85871147bb7c03c544fbc9c6515048ec470844d/vendor/github.com/openshift/client-go/build/clientset/versioned/typed/build/v1/buildconfig.go#L183-L196)

And it would still handle the ImageChangeTrigger tracking in buildConfig.Status that this PR and BUILD-186 is for.

Then, for the short term of the 4.8 release, where we still maintain the lastTriggeredImageID in the build config spec (i.e. https://github.com/openshift/openshift-apiserver/blob/f85871147bb7c03c544fbc9c6515048ec470844d/vendor/github.com/openshift/api/build/v1/types.go#L994) we easily can move that to the buildconfig controller in OCM, as updates to bc.Status.LastVersion in the apiserver will result in events arriving to the buildconfig controller  The build config controller can transfer lastTriggeredImageID's from Status to spec.  We can then remove that status to spec copying code in the OCM in 4.9

Would such a change reduce in any way @deads2k the pause you initially experience when you heard the apiserver was updating the buildconfig on its instantiate endpoint?

If so, at first blush I at least feel comfortable with such a change as part of BUILD-186

Option 2:
Moving out of the apiserver and into the OCM all buildconfig updates stemming from instantiating a build from the buildconfig's instantiate endpoint.

In thinking how this might be done, I could only come up with 

Option 2a: The build controller, as part of watching build objects, needs to update bc.Status.LastVersion when new builds come in, where based on the build name, it updates bc.Status.LastVersion, and updates bc.Status.ImageChangeTrigger[i].LastTriggeredImageID based on the triggerCause set on the build during build config instantiate.

Are we OK with the build controller updating build configs?  It is not currently.

Does this widen the timing window between when a build is instantiated from the build config instantiation endpoint and when bc.Status.LastVersion gets updated?  If so, does the possibility of generating new builds with the same name increase?

Similar questions for having mulitiple ICTs firing concurretly/simultaneously, resulting in more builds generated, if we update them in OCM vs. the apiserver?

Summary: not a super big code change, but I am not as confident about the timing impacts, etc.

thoughts @bparees @deads2k @adambkaplan ?

Option 2b: If we want the build controller to only update builds, and not build configs, then other than the relist on the buildconfig controller, I don't see how we initiate updates to the status LastVersion etc.  

A very frequent relist seems untenable to me.

adambkaplan commented:
> Option 2a: The build controller, as part of watching build objects, needs to update bc.Status.LastVersion when new builds come in, where based on the build name, it updates bc.Status.LastVersion, and updates bc.Status.ImageChangeTrigger[i].LastTriggeredImageID based on the triggerCause set on the build during build config instantiate.
> 
> Are we OK with the build controller updating build configs? It is not currently.

Isn't this contradictory? I thought the build controller _does_ update the buildConfig status today, when it sees that a build has reached a finished/completed state. So from an RBAC/permissions perspective, nothing changes.

> "if it ain't broke, don't fix it" is my view. The fact that that code has survived largely unchanged for 6 years says to me it's not something that needs to be rewritten.

It is also important to note that the scope of the issue is to merely _record_ the last image change ID in status. We are still using the information in `spec` to trigger builds. We plan to ignore the data in `spec` in OCP 4.9 - IMO this is right time to revisit the component that writes the last triggered image ID to the BuildConfig status.

gabemontero commented:
> > Option 2a: The build controller, as part of watching build objects, needs to update bc.Status.LastVersion when new builds come in, where based on the build name, it updates bc.Status.LastVersion, and updates bc.Status.ImageChangeTrigger[i].LastTriggeredImageID based on the triggerCause set on the build during build config instantiate.
> > Are we OK with the build controller updating build configs? It is not currently.
> 
> Isn't this contradictory? I thought the build controller _does_ update the buildConfig status today, when it sees that a build has reached a finished/completed state. So from an RBAC/permissions perspective, nothing changes.

Ah, I see now where my search on calls to UpdateStatus gave me an incomplete view @adambkaplan  - thanks.

I forgot, what the build controller does is enqueue events onto the buildconfig controller's workqueue when builds reach terminal state or are deleted.  For example, https://github.com/openshift/openshift-controller-manager/blob/0158a49aeb5696c3bd328eac51d6d5eb09d034c7/pkg/build/controller/build/build_controller.go#L1559
is the build terminal state / complete one.

In theory we could do the same when the build controller sees a new build, and that would trigger the buildconfig controller reconciling, updating status as needed.  We do not need to wait for something updating the buildconfig via an http post, etc.

> 
> > "if it ain't broke, don't fix it" is my view. The fact that that code has survived largely unchanged for 6 years says to me it's not something that needs to be rewritten.
> 
> It is also important to note that the scope of the issue is to merely _record_ the last image change ID in status. We are still using the information in `spec` to trigger builds. We plan to ignore the data in `spec` in OCP 4.9 - IMO this is right time to revisit the component that writes the last triggered image ID to the BuildConfig status.

Yeah at this point, you @adambkaplan and @bparees have strong opinions against updating now with what we know.  I'm at most not inclined to take a hard stance.  Assuming we want to reach an agreement with @deads2k on how to move forward, to verify whether or not keeping this as is a blocker for him, and if it is, why, we wait until he comes back from PTO on April 12, or we find an apiserver team member to act as proxy here.  I'm inclined to just wait for him.

deads2k commented on build/v1/types.go:
```diff
@@ -1022,23 +1026,48 @@ type ImageChangeTrigger struct {
-
-	// lastTriggerTime is the last time the BuildConfig was triggered by a change in the ImageStreamTag associated with this trigger.
-	LastTriggerTime metav1.Time `json:"lastTriggerTime,omitempty" protobuf:"bytes,4,opt,name=lastTriggerTime"`
+	From *ImageStreamTagReference `json:"from,omitempty" protobuf:"bytes,2,opt,name=from"`
```
Under what conditions is this information not known, but we were able to determine a SHA?  A SHA without telling me which imagestreamtagreference doesn't seem very helpful to general consumers.  It seems like we'd want to optimize for general consumers and if we own code that special cases the buildconfigfrom, it could actually know the value of the imagestream tag.

If for some reason taht isn't practical, indicating what it means when this is nil and how to actually find an imagestreamtag for it seems helpful.

gabemontero replied:
Yeah this should not be a pointer.  Obvious in hindsight.  I'll update in next push.  Thanks.

gabemontero replied:
to clarify a bit further, the current under the covers behavior where if the spec is nil, they fill in the strategy's from, as the godoc here cites, will be explicit here, meaning the strategy from values will exist in the corresponding entry for it in our new array in the bc.status for ICTs

deads2k commented:
I'd like to see https://github.com/openshift/api/pull/882/files#r611900039 addressed one way or the other, but the remaining updates look good.

/approve

gabemontero commented:
OK @bparees @adambkaplan I've
- squashed commits from earlier reviews, including some error checking / tweaks in the godoc I did today
- addressed @deads2k 's https://github.com/openshift/api/pull/882/files#r611900039  (new field does not need to be  pointer ... godoc explains an entry can come from the build strategy from)

PTAL - thanks

deads2k commented:
/lgtm
/approve

adambkaplan commented on build/v1/types.go:
```diff
@@ -1022,23 +1026,51 @@ type ImageChangeTrigger struct {
 	Paused bool `json:"paused,omitempty" protobuf:"varint,3,opt,name=paused"`
 }
 
+// ImageStreamTagReference captures the required elements for identifying the ImageStreamTag referenced by the more
+// generic ObjectReference BuildTriggerPolicy.ImageChange.From.  It is used by ImageChangeTriggerStatus, where a
+// specific instance of ImageChangeTriggerStatus in maintained in BuildConfigStatus.ImageChangeTriggers for each entry
+// in the BuildConfigSpec.Triggers array where the BuildTriggerPolicy.ImageChange pointer is set to a non-nil value
```
Likewise, too many golang details for API documentation. The gory details can live here in the PR and any git commits.

```suggestion
// ImageStreamTagReference references the ImageStreamTag used in an image change trigger by namespace and name.
```

adambkaplan commented on build/v1/types.go:
```diff
@@ -1022,23 +1026,51 @@ type ImageChangeTrigger struct {
-	// lastTriggeredImageID is the sha/id of the imageref cited in the 'from' field the last time this BuildConfig was triggered.
-	// It is not necessarily the sha/id of the image change that triggered a build.
-	// This field is updated for all image change triggers when any of them triggers a build.
+	// lastTriggeredImageID represents, at the last time a Build for this BuildConfig was instantiated, the sha/id of
+	// the image referenced by the the ImageStreamTag cited in the 'from' of this struct.
+	// The lastTriggeredImageID field will be updated by the OpenShift APIServer on all instantiations of a Build from
+	// the BuildConfig it processes, regardless of what is considered the cause of instantiation.
+	// Specifically, an instantiation of a Build could have been manually requested, or could have resulted from
+	// changes with any of the Triggers defined in BuildConfigSpec.Triggers.
+	// The reason for always updating this field across all ImageChangeTriggerStatus instances is to prevent
+	// multiple builds being instantiated concurrently when multiple ImageChangeTriggers fire concurrently.  The system
+	// compares the the sha/id stored here with the associated ImageStreamTag's sha/id for the image.  If they match,
+	// then this trigger is not a valid reason for instantiating a Build.  So when ImageChangeTriggers fire concurrently,
+	// only one of them can "win", meaning selected as the cause for a Build instantiation request.
+	// Lastly, to clarify exactly what is meant by "Build instantiation", from a REST perspective, it is a HTTP POST of a
+	// BuildRequest object as the HTTP Body that is made to the OpenShift APIServer, where that HTTP POST also specifies
+	// the "buildconfigs" resource,  "instantiate" subresource, as well as the namespace and name of the BuildConfig.
```
Likewise, too many details for API documentation.

```suggestion
	// lastTriggeredImageID represents the sha/id of the ImageStreamTag used when a Build for this BuildConfig was started.
	// The lastTriggeredImageID will be updated each time a Build for this BuildConfig is started, even if this ImageStreamTag is not the reason the Build is started.
```

gabemontero commented:
@bparees assuming you are good with these changes as well, this is another pillar lead apply the valid bz label for non feature freeze etc etc .... thanks

adambkaplan commented on build/v1/types.go:
```diff
@@ -1022,23 +1026,51 @@ type ImageChangeTrigger struct {
+	// ImageChange Trigger in the spec with a 'nil' for its 'from'.
+	From ImageStreamTagReference `json:"from,omitempty" protobuf:"bytes,2,opt,name=from"`
+
+	// lastTriggerTime is the last time this particular ImageChangeTrigger fired, and that trigger firing was chosen as the cause for the Build being instantiated
+	// from this BuildConfig.  So on each Build instantiation, while lastTriggeredImageID will be updated regardless of
+	// whether this ImageChangeTrigger fired and deemed the cause for the Build Instantiation, this field is only updated
+	// when this trigger was in fact deemed the cause.  As such, it is valid that this field may not be set across all the
+	// ImageChangeTriggers, as they may have not yet been deemed to be the cause of a Build instantiation.  It is also
+	// valid that the times stored in lastTriggerTime will vary across all the ImageChangeTriggers, as the system
+	// explicitly picks only one trigger cause for a given Build.
```
```suggestion
	// lastTriggerTime is the last time this particular ImageStreamTag triggered a Build to start.
	// This field is only updated when this trigger specifically started a Build.
```

adambkaplan commented on build/v1/types.go:
```diff
@@ -1022,23 +1026,51 @@ type ImageChangeTrigger struct {
-
-	// lastTriggerTime is the last time the BuildConfig was triggered by a change in the ImageStreamTag associated with this trigger.
-	LastTriggerTime metav1.Time `json:"lastTriggerTime,omitempty" protobuf:"bytes,4,opt,name=lastTriggerTime"`
+	// This can come from an ImageStream tag referenced in this BuildConfig's Spec ImageChange Triggers, or the "from"
+	//  this BuildConfig's build strategy if it happens to be an ImageStreamTag (where the user has specified an
+	// ImageChange Trigger in the spec with a 'nil' for its 'from'.
```
I think these comments can be removed. From an API perspective, the first sentence captures everything an end user needs to know.

adambkaplan reviewed (changes requested):
API doc comments need to be cleaned up, as they ultimately are end user documentation. I may want another set of eyes to ensure grammar+voice are consistent.

/cc @rolfedh

rolfedh commented:
> API doc comments need to be cleaned up, as they ultimately are end user documentation. I may want another set of eyes to ensure grammar+voice are consistent.
> 
> /cc @rolfedh

ack

rolfedh commented on build/v1/types.go:
```diff
@@ -977,7 +977,11 @@ type BuildConfigStatus struct {
 	LastVersion int64 `json:"lastVersion" protobuf:"varint,1,opt,name=lastVersion"`
 
 	// ImageChangeTriggers is used to capture the runtime state of any ImageChangeTrigger specified in the BuildConfigSpec,
```
// ImageChangeTriggers captures the runtime state of any ImageChangeTrigger specified in the BuildConfigSpec

rolfedh commented on build/v1/zz_generated.swagger_doc_generated.go:
```diff
@@ -101,7 +101,7 @@ func (BuildConfigSpec) SwaggerDoc() map[string]string {
 	"":                    "BuildConfigStatus contains current state of the build config object.",
 	"lastVersion":         "lastVersion is used to inform about number of last triggered build.",
-	"imageChangeTriggers": "ImageChangeTriggers is used to capture the runtime state of any ImageChangeTrigger specified in the BuildConfigSpec, including reconciled values of the lastTriggeredImageID and paused fields.",
+	"imageChangeTriggers": "ImageChangeTriggers is used to capture the runtime state of any ImageChangeTrigger specified in the BuildConfigSpec, including the value reconciled by the OpenShift APIServer for the lastTriggeredImageID.  There will be a single entry in this array for each entry in the BuildConfigSpec.Triggers array where the BuildTriggerPolicy.ImageChange pointer is set to a non-nil val ...
```
Will this be regenerated to reflect the preceding changes?

adambkaplan replied:
yes, but we need to do it in a separate pull request (current code has merged).

rolfedh commented:
Everything looks good. Although widely used throughout openshift/api, if they don't change or weaken the meaning, I recommend replacing:

- future-tense with present tense. For example s/will be/is/ 
- "is used to" with the action. For example, s/is used to capture/captures/
- `.  ` with `. `
